Escape like parameters in database user backend

LukasReschke opened the pull request:
This leads to unexpected behaviour otherwise and pretty sure also some very confusing bugs.

@icewind1991 Objections?
@daita Should also fix your issue :)

LukasReschke committed:
Escape like parameters in database user backend

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke commented on lib/private/User/Database.php:
```diff
@@ -76,6 +79,7 @@ class Database extends Backend implements IUserBackend {
 	public function __construct($eventDispatcher = null) {
 		$this->cache = new CappedMemoryCache();
 		$this->eventDispatcher = $eventDispatcher ? $eventDispatcher : \OC::$server->getEventDispatcher();
+		$this->connection = \OC::$server->getDatabaseConnection();
```
Let's see if we can move that to DI. Want @icewind1991's feedback however first if this is the right approach.

LukasReschke replied:
I bet however that the dependency is too early, thus also the failing tests. But if I get told that this is the right function to use I can adjust our code to work around that :)

nickvergessen replied:
Yes function usage looks good

daita commented:
to anyone wondering: 

> 21:06 <@maxence> is it a known feature that when you type the name of a user/group when sharing, you can 'search' by replacing any char by an underscore ?
> 21:06 <@maxence> you can find the user 'test' by writing t__t
>

LukasReschke committed:
Don't call function in constructor

The constructor is iniitiated already very early in base.php, thus requiring this here will break the setup and some more. For now we probably have to live with a static function call here thus.

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke commented:
DI doesn't really seem to work here but at least the tests should run now 😄
